sha256-avx2/utilsx8.c

```diff
@@ -5,121 +5,142 @@
-                   stackx8 + 5*(tree_height + 1)*SPX_N + offset*SPX_N,
-                   stackx8 + 6*(tree_height + 1)*SPX_N + offset*SPX_N,
-                   stackx8 + 7*(tree_height + 1)*SPX_N + offset*SPX_N,
-                   sk_seed, pub_seed,
-                   idx + idx_offset[0],
-                   idx + idx_offset[1],
-                   idx + idx_offset[2],
-                   idx + idx_offset[3],
-                   idx + idx_offset[4],
-                   idx + idx_offset[5],
-                   idx + idx_offset[6],
-                   idx + idx_offset[7],
-                   tree_addrx8);
-        offset++;
-        heights[offset - 1] = 0;
+        /* Now combine the freshly generated right node with previously */
+        /* generated left ones */
+        uint32_t internal_idx_offset = idx_offset;
+        uint32_t internal_idx = idx;
+        uint32_t internal_leaf = leaf_idx;
+        uint32_t h;     /* The height we are in the Merkle tree */
+        for (h=0;; h++, internal_idx >>= 1, internal_leaf >>= 1) {
 
-        /* If this is a node we need for the auth path.. */
-        for (j = 0; j < 8; j++) {
-            if ((leaf_idx[j] ^ 0x1) == idx) {
-                memcpy(auth_pathx8 + j*tree_height*SPX_N,
-                       stackx8 + j*(tree_height + 1)*SPX_N + (offset - 1)*SPX_N, SPX_N);
+            /* Special processing if we're at the top of the tree */
+            if (h >= tree_height - 3) {
+                if (h == tree_height) {
+                    /* We hit the root; return it */
+                    memcpy( root, &current[7*SPX_N], SPX_N );
+                    return;
+                }
+                /* The tree indexing logic is a bit off in this case */
+                /* Adjust it so that the left-most node of the part of */
+                /* the tree that we're processing has index 0 */
+                prev_left_adj = left_adj;
+                left_adj = 8 - (1 << (tree_height - h - 1));
             }
-        }
 
-        /* While the top-most nodes are of equal height.. */
-        while (offset >= 2 && heights[offset - 1] == heights[offset - 2]) {
-            /* Compute index of the new node, in the next layer. */
-            tree_idx = (idx >> (heights[offset - 1] + 1));
+            /* Check if we hit the top of the tree */
+            if (h == tree_height) {
+                /* We hit the root; return it */
+                memcpy( root, &current[7*SPX_N], SPX_N );
+                return;
+            }
+            
+            /*
+             * Check if one of the nodes we have is a part of the
+             * authentication path; if it is, write it out
+             */
+            if ((((internal_idx << 3) ^ internal_leaf) & ~0x7) == 0) {
+                memcpy( &auth_path[ h * SPX_N ],
+                        &current[(((internal_leaf&7)^1) + prev_left_adj) * SPX_N],
+                        SPX_N );
```

Comment:
This line also deserves some comments. We know that one of our eight nodes at hand should go into the authentication path.   The node that should go in the authentication path is given by `internal_leaf ^1` (the sibling of the node in the signature path.)  If we still have eight nodes at hand, then the index of the first of them is a multiple of eight, so we can simply read of the index of that node in the `current` array by reading the lower 3 bits.

However, in the last three levels the nodes are shifted to the right in the `current` array which is tracked by `left_adj` and `prev_left_adj`.  As we're interested in a value generated in the previous level, we need to use `prev_left_adj`.